tests/api2/test_999_pool_dataset_unlock.py

```diff
@@ -0,0 +1,178 @@
+
+
+def passphrase_encryption():
+    return {
+        'encryption_options': {
+            'generate_key': False,
+            'pbkdf2iters': 100000,
+            'algorithm': 'AES-128-CCM',
+            'passphrase': 'passphrase',
+        },
+        'encryption': True,
+        'inherit_encryption': False,
+    }
+
+
+@contextlib.contextmanager
+def dataset(name, options=None):
+    assert "/" not in name
+
+    dataset = f"{pool_name}/{name}"
+
+    result = POST("/pool/dataset/", {"name": dataset, **(options or {})})
+    assert result.status_code == 200, result.text
+
+    try:
+        yield dataset
+    finally:
+        result = DELETE(f"/pool/dataset/id/{urllib.parse.quote(dataset, '')}/")
+        assert result.status_code == 200, result.text
+
+
+@contextlib.contextmanager
+def smb_share(name, path, options=None):
+    results = POST("/sharing/smb/", {
+        "name": name,
+        "path": path,
+        "guestok": True,
+        **(options or {}),
+    })
+    assert results.status_code == 200, results.text
+    id = results.json()["id"]
+
+    try:
+        yield id
+    finally:
+        result = DELETE(f"/sharing/smb/id/{id}/")
+        assert result.status_code == 200, result.text
+
+
+def lock_dataset(name):
+    payload = {
+        'id': name,
+        'lock_options': {
+            'force_umount': True
+        }
+    }
+    results = POST('/pool/dataset/lock', payload)
+    assert results.status_code == 200, results.text
+    job_id = results.json()
+    job_status = wait_on_job(job_id, 120)
+    assert job_status['state'] == 'SUCCESS', str(job_status['results'])
+
+
+def unlock_dataset(name, options=None):
+    payload = {
+        'id': name,
+        'unlock_options': {
+            'recursive': True,
+            'datasets': [
+                {
+                    'name': name,
+                    'passphrase': 'passphrase'
+                }
+            ],
+            **(options or {}),
+        }
+    }
+    results = POST('/pool/dataset/unlock/', payload)
+    assert results.status_code == 200, results.text
+    job_id = results.json()
+    job_status = wait_on_job(job_id, 120)
+    assert job_status['state'] == 'SUCCESS', str(job_status['results'])
+    assert job_status['results']['result']['unlocked'] == [name], str(job_status['results'])
+
+
+def run(command):
+    return subprocess.run(command, shell=True, check=True, capture_output=True, encoding="utf-8")
+
+
+@contextlib.contextmanager
+def smb_connection(**kwargs):
+    c = SMB()
+    c.connect(**kwargs)
+
+    try:
+        yield c
+    finally:
+        c.disconnect()
+
+
+@pytest.mark.dependency(name="create_dataset")
+@pytest.mark.parametrize("attachments_list_mode", ["ALLOW", "DENY"])
+def test_pool_dataset_unlock_smb(request, attachments_list_mode):
+    #depends(request, ["pool_04", "smb_001"], scope="session")
+    # Prepare test SMB share
+    with dataset("normal") as normal:
+        os.chmod(f"/mnt/{normal}", 0o777)
```

Comment:
This test will be run on an external server, no?
`POST('/filesystem/setperm/', {'path': f'/mnt/{normal}', 'mode': "777"})`